<a href="https://colab.research.google.com/github/Martinez-Ortiz-Zaira-Natalia/M-todos-Num-ricos-1/blob/main/POLINOMIO%20DE%20%20HERMITE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Polinomio de Hermite


In [ ]:
#Definir la biblioteca  que usaremos
import numpy as np

In [ ]:
# Guardar los datos a usar
x = np.array([1.3, 1.6, 1.9])  # Aqui puedes editar los datos
f = np.array([0.6200860, 0.4554022, 0.2818186])  #  valores de f(x)
f_prime = np.array([-0.5220232, -0.5698959, -0.5811571])  # derivadas de f(x)

In [ ]:
# Solicitar al usuario que ingrese el valor de x_eval
y = float(input("En qué punto deseas evaluar? "))

En qué punto deseas evaluar? 1.5


In [ ]:
# Paso 1:
def hermite_interpolation(x, f, f_prime, x_eval):
    n = len(x)  # Numero de puntos
    z = np.zeros(2 * n)  # Z sera el arreglo que contiene los nodos duplicados
    Q = np.zeros((2 * n, 2 * n))  # La tabla de diferencias divididas

    # Paso 2: Duplicar los puntos en z y asignar los valores de la funcion y su derivada
    for i in range(n):
        z[2 * i] = x[i]
        z[2 * i + 1] = x[i]
        Q[2 * i, 0] = f[i]
        Q[2 * i + 1, 0] = f[i]
        Q[2 * i + 1, 1] = f_prime[i]

    # Paso 3: Calcular las primeras diferencias divididas para el caso i distinto de 0
    for i in range(1, n):
        Q[2 * i, 1] = (Q[2 * i, 0] - Q[2 * i - 1, 0]) / (z[2 * i] - z[2 * i - 1])

    # Paso 4: Calcular las diferencias divididas para i = 2, 3,..., 2n + 1
    for i in range(2, 2 * n):
        for j in range(2, i + 1):
            Q[i, j] = (Q[i, j - 1] - Q[i - 1, j - 1]) / (z[i] - z[i - j])

    # Imprime la tabla de diferencias divididas
    print("Tabla de diferencias divididas (Con respecto a Hermite):")

    # Definir los encabezados de las columnas de manera fija para los primeros 2 columnas
    headers = ["x"] + [f"Q[i][{i}]" for i in range(0, 2 * n)]

    # Imprime los encabezados de la tabla
    print("          ".join(headers))  # Imprime los titulos de las columnas

    for i in range(2 * n):
        # Imprime x y las diferencias divididas Q[i][j]
        print(f"{z[i]:.6f}", end="\t")  # Muestra el valor de x
        for j in range(i + 1):
            print(f"{Q[i, j]:.6f}", end="\t")  # Muestra los valores de las diferencias divididas
        print()  # Nueva línea para cada fila

    # Paso 5: Evaluar el polinomio de Hermite  Donde da la Salida
    result = Q[0, 0]  # El primer valor es f(z_0)
    product = 1  # Empezamos con el primer término como 1 (es (x_eval - z_0)^0)

     # Generar el polinomio como una cadena de texto
    poly_terms = [f"{Q[0, 0]:.6f}"]

    # Generar los terminos del polinomio de Hermite
    for i in range(1, 2 * n):
        term = f"(x - {z[i - 1]:.6f})"
        for j in range(1, i):
            term += f" * (x - {z[j - 1]:.6f})"
        term += f" * {Q[i, i]:.6f}"
        poly_terms.append(term)

    # Imprimir el polinomio completo
    print("\nEl polinomio de Hermite generado es:")
    poly_string = " + ".join(poly_terms)
    print(poly_string)

    # Evalua el polinomio en el valor x_eval
    evaluation_result = Q[0, 0]
    product = 1
    for i in range(1, 2 * n):
        product *= (x_eval - z[i - 1])
        evaluation_result += Q[i, i] * product

    return evaluation_result

In [ ]:
#Guarda el valor del punto x
x_eval=y
result = hermite_interpolation(x, f, f_prime, x_eval)
# Imprime el resultado de la evaluacion
print(f"\nValor del polinomio en x es {x_eval} : {result:.6f}")

Tabla de diferencias divididas (Con respecto a Hermite):
x          Q[i][0]          Q[i][1]          Q[i][2]          Q[i][3]          Q[i][4]          Q[i][5]
1.300000	0.620086	
1.300000	0.620086	-0.522023	
1.600000	0.455402	-0.548946	-0.089743	
1.600000	0.455402	-0.569896	-0.069833	0.066366	
1.900000	0.281819	-0.578612	-0.029054	0.067966	0.002667	
1.900000	0.281819	-0.581157	-0.008484	0.068567	0.001002	-0.002775	

El polinomio de Hermite generado es:
0.620086 + (x - 1.300000) * -0.522023 + (x - 1.300000) * (x - 1.300000) * -0.089743 + (x - 1.600000) * (x - 1.300000) * (x - 1.300000) * 0.066366 + (x - 1.600000) * (x - 1.300000) * (x - 1.300000) * (x - 1.600000) * 0.002667 + (x - 1.900000) * (x - 1.300000) * (x - 1.300000) * (x - 1.600000) * (x - 1.600000) * -0.002775

Valor del polinomio en x es 1.5 : 0.511828


En este programa implementamos  el método de diferencias divididas para calcular el polinomio osculante de Hermite del ejercicio 1 de la página 101, libro Burden.

\begin{array}{|c|c|c|c|}
\hline
k & x_k & f(x_k) & f'(x_k) \\
\hline
0 & 1.3 & 0.6200860 & -0.5220232 \\
1 & 1.6 & 0.4554022 & -0.5698959 \\
2 & 1.9 & 0.2818186 & -0.5811571 \\
\hline
\end{array}

Además, evaluamos en un valor de x en este caso $x=1.5$

